saw-remote-api/src/SAWServer/Data/MIRType.hs

```diff
@@ -54,36 +60,76 @@ instance JSON.FromJSON MIRTypeTag where
       "usize" -> pure TagUsize
       _ -> empty
 
-newtype JSONMIRType = JSONMIRType { getMIRType :: Ty }
+-- | This is like 'Mir.Ty', but with the following differences:
+--
+-- 1. This only contains the subset of MIR types that are currently supported
+--    by the SAW remote API.
+--
+-- 2. 'JSONTyAdt' only contains a 'ServerName' that points points to an
+--    'Mir.Adt', as opposed to 'Mir.TyAdt', which contains a full 'Mir.Adt'.
+--    The advantage of only containing a 'Mir.TyAdt' is that we do not have to
+--    represent the entirety of a 'Mir.Adt' definition in JSON each time we want
+--    to reference the ADT in a type.
+data JSONMIRType
+  = JSONTyAdt !ServerName
+  | JSONTyArray !JSONMIRType !Int
+  | JSONTyBool
+  | JSONTyFloat !Mir.FloatKind
+  | JSONTyInt !Mir.BaseSize
+  | JSONTySlice !JSONMIRType
+  | JSONTyStr
+  | JSONTyTuple ![JSONMIRType]
+  | JSONTyUint !Mir.BaseSize
```

Comment:
I changed `JSONMIRType` from a newtype around `Mir.Ty` to a distinct data type solely for the purpose of making the `Adt` case (`JSONMIRType`) store a `ServerName` pointing to a `Mir.Adt` rather than storing a `Mir.Adt` directly. Storing a `Mir.Adt` directly would have meant that we would have had to store the entire `Mir.Adt` as JSON, which would have been quite expensive, as `Mir.Adt` is a large data structure. Moreover, it's unnecessary to do so, since the ADT definition is already contained in the MIR module, so storing the `Mir.Adt` would be redundant.